src/vm/assemblyspec.cpp

```diff
@@ -1660,9 +1672,29 @@ BOOL AssemblySpecBindingCache::StoreFile(AssemblySpec *pSpec, PEAssembly *pFile)
     if (entry == (AssemblyBinding *) INVALIDENTRY)
     {
         AssemblyBindingHolder abHolder;
-        entry = abHolder.CreateAssemblyBinding(m_pHeap);
 
-        entry->Init(pSpec,pFile,NULL,NULL,m_pHeap, abHolder.GetPamTracker());
+        LoaderHeap* pHeap = m_pHeap;
+
+#ifndef CROSSGEN_COMPILE
+        if (pBinderContextForLookup != NULL)
+        {
+            LoaderAllocator* pLoaderAllocator = NULL;
+
+            // Assemblies loaded with AssemblyLoadContext need to use a different heap if
+            // marked as collectible
+            SafeComHolder<ICollectibleAssemblyLoadContext> pAssemblyLoadContext = NULL;
+            if (SUCCEEDED(pBinderContextForLookup->QueryInterface<ICollectibleAssemblyLoadContext>(&pAssemblyLoadContext)))
+            {
+                pAssemblyLoadContext->GetLoaderAllocator(&pLoaderAllocator);
+                _ASSERTE(pLoaderAllocator != NULL);
+                pHeap = pLoaderAllocator->GetHighFrequencyHeap();
+            }
+        }
+#endif // !CROSSGEN_COMPILE
+
```

Comment:
These changes are effectively creating a binding from an assemblySpec which lives in one loaderallocator to the loaderallocator of the collectible assembly. I don't see the associated logic that would add the the loaderallocator graph so that we maintain lifetime correctly, in the presence of interrelated collectible assembly load contexts

Reply:
Similar to the AssemblySpecBindingCache::StoreAssembly, I've tried to modify the way we get the binder context to use the context from the pFile and not try to use the context from the pSpec. @davidwrighton, it seems this should remove the issue, right?

Reply:
? That sounds like a semantic change to assembly loading. The context that we use is a critical thing, and has implications for future binds. Lets talk about that.